### ISSUE TITLE
grafana login by oauth login.OAuthLogin(missing saved state)

### ISSUE THREAD

**Issue by xiaoxi1989** [grafana/grafana#17925]
**What happened**:
grafana auth by keycloak and session store in mysql.
first login get error: login.OAuthLogin(missing saved state),
![image](https://user-images.githubusercontent.com/23303904/60657611-3430dd00-9e84-11e9-8c79-929aeb92d1f6.png)

but relogin by **(sign in with oauth)** is fine(no input user and password).
![image](https://user-images.githubusercontent.com/23303904/60657658-427ef900-9e84-11e9-99f4-31ac6847fe27.png)

**What you expected to happen**:
login success

**Environment**:
- Grafana version:
grafana 6.25

- Data source type & version:
mysql5.7

- OS Grafana is installed on:
docker

- User OS & Browser:
macos 1.13.6 
chrome

**Comment by torkelo** [comment]
That is strange, we have no similar reports. Do you always get this on the first login?

**Comment by xiaoxi1989** [comment]
@torkelo  Thank you for your reply. yes, every first login get this error.
my configuration is 
```
[auth.generic_oauth]
enabled = true
name = OAuth
allow_sign_up = true
client_id = grafana
client_secret = 6886cf64-ab2f-42a6-8ef9-c43380d94230
scopes = user:email
auth_url = http://xxxx:18081/auth/realms/grafana/protocol/openid-connect/auth
token_url = http://xxxx:18081/auth/realms/grafana/protocol/openid-connect/token
api_url = http://xxxx:18081/auth/realms/grafana/protocol/openid-connect/userinfo

[database]
# You can configure the database connection by specifying type, host, name, user and password
# as separate properties or as on string using the url properties.
# Either "mysql", "postgres" or "sqlite3", it's your choice
type = mysql
host = xxxxx:3306
name = grafana
user = root
# If the password contains # or ; you have to wrap it with triple quotes. Ex """#password;"""
password = xxxxx


```

**Comment by xiaoxi1989** [comment]
request info
![image](https://user-images.githubusercontent.com/23303904/60704227-6864d600-9f36-11e9-8e2b-79713a2f4a4e.png)

**Comment by xiaoxi1989** [comment]
chrome info 
curl 'http://localhost:3000/login/generic_oauth?state=k8oYTIDNn14diTQ6tz0vQvrcYslEhf7c4cVlD3pSNK4%3D&session_state=2e079150-5234-47ef-9f3f-71da807edfdd&code=8f996c18-4420-4fde-86fd-446a65f65be1.2e079150-5234-47ef-9f3f-71da807edfdd.a37ecfc6-07eb-46ed-bd25-ae8d519cb43e' -H 'Connection: keep-alive' -H 'Cache-Control: max-age=0' -H 'Upgrade-Insecure-Requests: 1' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36' -H 'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3' -H 'Referer: http://10.13.83.10:18081/auth/realms/grafana/protocol/openid-connect/auth?access_type=online&client_id=grafana&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Flogin%2Fgeneric_oauth&response_type=code&scope=user%3Aemail&state=k8oYTIDNn14diTQ6tz0vQvrcYslEhf7c4cVlD3pSNK4%3D' -H 'Accept-Encoding: gzip, deflate, br' -H 'Accept-Language: zh-CN,zh;q=0.9' -H 'Cookie: oauth_state=ea46359a58f6951d7091c8e7def416f0e067ea272e4e66c416bb93192946f2af' --compressed

Response Cookies
oauth_state is empty

**Comment by xiaoxi1989** [comment]
this is my fault

i changed grafana.ini to solve the problem.
```
domain = 0.0.0.0
# Redirect to correct domain if host header does not match domain
# Prevents DNS rebinding attacks
enforce_domain = true
# The full public facing url you use in browser, used for redirects and emails
# If you use reverse proxy and sub path specify full url (with sub path)
root_url = http://0.0.0.0:3000
```

**Comment by haf** [comment]
I get the same error; but I'm reluctant to change to `domain = 0.0.0.0` since it's not a domain. Config:

```ini
instance_name = grafana.example.com

[server]
root_url = https://grafana.example.com
domain = grafana.example.com
enforce_domain = false
enable_gzip = true

[security]
admin_password = snip
cookie_secure = true
cookie_samesite = strict
auto_assign_org = true
auto_assign_org_id = 1
auto_assign_org_role = Viewer
viewers_can_edit = false
editors_can_admin = true # default false
login_remember_days = 30

[auth.github]
enabled = true
allow_sign_up = true
client_id = snip
client_secret = snip
scopes = user:email,read:org
auth_url = https://github.com/login/oauth/authorize
token_url = https://github.com/login/oauth/access_token
api_url = https://api.github.com/user

[analytics]
reporting_enabled = false
```

using k8s config:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: grafana
    component: grafana
    tier: resources
  name: grafana
  namespace: monitoring
spec:
  selector:
    matchLabels:
      app: grafana
      tier: resources
  strategy:
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
    type: RollingUpdate
  template:
    metadata:
      annotations:
        prometheus.io/path: /metrics
        prometheus.io/port: "3000"
        prometheus.io/scrape: "true"
      labels:
        app: grafana
        tier: resources
    spec:
      containers:
      - image: grafana/grafana:6.3.2
        name: grafana
        ports:
        - containerPort: 3000
          name: http
        resources:
          limits:
            cpu: 500m
            memory: 2500Mi
          requests:
            cpu: 100m
            memory: 100Mi
        volumeMounts:
        - mountPath: /var/lib/grafana
          name: data
        - mountPath: /etc/grafana/grafana.ini
          name: conf
          subPath: grafana.ini
      initContainers:
      - command:
        - chown
        - -R
        - 472:472
        - /var/lib/grafana
        image: busybox:1.30
        name: chown
        securityContext:
          runAsUser: 0
        volumeMounts:
        - mountPath: /var/lib/grafana
          name: data
      securityContext:
        fsGroup: 472
        runAsUser: 472
      volumes:
      - name: data
        persistentVolumeClaim:
          claimName: grafana
      - configMap:
          name: grafana
        name: conf
```

With a HTTPS-terminating nginx proxy in front.

**Comment by marefr** [comment]
@haf what browser are you using? Chrome? Safari? Firefox?

If you check the 302 response of the request "github", 2 steps up from the highlighted one, does it have a set-cookie header and if so, what's the content of that one?

**Comment by haf** [comment]
@marefr I tested both in Firefox and Chromium. No, it's a brand new install, with the above config.

Yes, it has a `set-cookie` header; it looks like this;

<img width="651" alt="Screenshot 2019-08-07 at 16 13 50" src="https://user-images.githubusercontent.com/193115/62642110-57e4b880-b945-11e9-950a-89a31e755175.png">

```
oauth_state=3d3c1e0648a6fd630ef4ad2SNIP75f165e80d1ed4992e4; Path=/; Max-Age=60; HttpOnly; Secure; SameSite=Strict
```

**Comment by haf** [comment]
`Set-Cookie` does not work when `SameSite=Strict` and there's a third party domain in between https://stackoverflow.com/questions/42216700/how-can-i-redirect-after-oauth2-with-samesite-strict-and-still-get-my-cookies/42220786

**Comment by marefr** [comment]
@haf Thanks.

`cookie_samesite = lax` is the default for a reason.

**Comment by haf** [comment]
It's inviting to change to strict, since that somewhat improves the security stance; it's also a hidden hole you can fall into; wouldn't it be better to use an alternative state storage mechanism as a bug fix, to avoid confusing users and making it possible to use federated authentication with strict samesite-policies?

**Comment by marefr** [comment]
@haf Yes, absolutely. This was how it worked before when we had session storage, but we had to switch to cookie when removing session storage since we didn't have another option at that time. I will open a new issue to track this change and reference this issue. Thanks

Glad to see you checking in here btw, long time, no see

**Comment by haf** [comment]
@marefr Nice to see you too Marcus :). We should grab a beer at some point during the fall!

**Comment by gregparsons** [comment]
Is this a 2020 version of the original error?

![sadly](https://user-images.githubusercontent.com/4780760/75600473-8d2d3100-5a64-11ea-979e-532c66176763.jpg)

login.OAuthLogin(missing saved state)

I'm on 6.6.2 using effectively the same config, Okta, cookie_samesite = lax, Nginx reverse proxy, and the domain set in GF_SERVER_ROOT_URL. Changing to GF_SECURITY_COOKIE_SAMESITE=none results in first-ever openid logins continuing correctly. I'm not sad, but am I kicking a can down the road? Will [18456](https://github.com/grafana/grafana/issues/18456#issue-478407144) be relevant?

**Comment by carllacandazo** [comment]
I have the same error and setting my  cookie_samesite = lax is not working for me, may I know another solution for this? Thanks

**Comment by marefr** [comment]
We closed #18456 since we need to use a cookie as a security measure to be sure request did originate from Grafana and no one have tampered with it.

Make sure to read details in https://grafana.com/docs/grafana/latest/installation/upgrading/#important-changes-regarding-samesite-cookie-attribute 

I would suggest using chrome developer tools or similar and inspect network tab, check Preserve log and do the oauth login and inspect cookies being set on each Grafana and auth provider request to backtrack where things go wrong.

**Comment by carllacandazo** [comment]
Is there a known issue when using Grafana oauth in Incognito?

I'm currently using Chrome Incognito to test the Grafana and here is the cookie error:
![image](https://user-images.githubusercontent.com/34564514/79209130-96882680-7e75-11ea-82e7-f16089dbd73b.png)

**Comment by marefr** [comment]
@carllacandazo please inspect cookies and Set-Cookie header on the request redirecting from Grafana to your oauth provider.

**Comment by carllacandazo** [comment]
I think the Set-Cookie is working, also i've double clicked the generic_oauth and it's redirecting me to the correct URL and i'm successfully authenticated.

![image](https://user-images.githubusercontent.com/34564514/79351899-cdd8ff00-7f6b-11ea-9878-fdb9cd387d9f.png)

**Comment by varlen** [comment]
I had the same problem. Inspecting Grafana logs I realized that the user didn't had an email setup on my OAuth provider. Adding the email address solved the problem.

**Comment by carllacandazo** [comment]
@varlen can you please show the logs?

**Comment by varlen** [comment]
> @varlen can you please show the logs?

I don't have the logs anymore but I tried to reproduce this behavior and Grafana gave me a different error... So now I'm not so sure what the cause was anymore.

**Comment by Reforced** [comment]
I've just had this same issue, don't know what caused the "saved state" to be gone. The solution was simple luckily; manually deleting the current cookies in my browser fixed it. (Chrome 81.0.4044.138)

**Comment by noctarius** [comment]
I have the same issue, but only with my mobile phone. All other logins (also same account on a laptop) work fine.

Mobile phone unfortunately means no developer tools. Phone is Oneplus 7, with Chrome browser. Auth provider is Google. In general, as mentioned, logins work flawlessly, just not on the phone. And even that just started to happen on morning out of the blue. 

Tried removing all settings and files for our side in Chrome (mobile). Is there another domain to remove?

**Comment by gregparsons** [comment]
It's back. But only on Chrome. Anybody see the same using Okta? On MacOS and Windows. The site works fine on every other browser, mobile or otherwise. The cookie isn't getting set. 

<img width="370" alt="image" src="https://user-images.githubusercontent.com/4780760/90530087-54c00300-e129-11ea-9e2c-eb9da083ae42.png">

**Comment by gsaileshkumar** [comment]
I use Auth0 as oauth provider and face this issue quite frequently. However on analysis found out the root cause for my issue is as below,
Grafana version: 7.0.3
Relevant configs for this issue:
GF_AUTH_OAUTH_AUTO_LOGIN=true
GF_AUTH_GENERIC_OAUTH_ENABLED=true
GF_AUTH_SIGNOUT_REDIRECT_URL=<Auth0_domain>/logout?returnTo=<login_page>

As far as I understand, every time Grafana lands on login page, following actions occur
- Api call to {application_domain}/login 
  - This sets `oauth_state` cookie with expiry of 60 seconds (default)
- Redirects to {application_domain}/login/generic_auth (Since auto login via oauth is enabled)
- Redirects to Auth0 endpoints to perform authentication
  - {Auth0_domain}/authorize
  - {Auth0_domain}/login - returns Auth0 login page
- User enters login/password and clicks enter. Following api calls are made by Auth0 for authentication
  - {Auth0_domain}/login 
  - {Auth0_domain}/callback
- Redirects to {application_domain}/login/generic_auth on successful authentication
  - Grafana checks for the `oauth_state` cookie which it set in step 1
Now following checks are done in Grafana
- `oauth_state` Cookie is present
  - Sets `grafana_session` cookie
  - If cookie is not present `login.OAuthLogin(missing saved state)` error is returned
  - If present, continue
- `oauth_state` Cookie is valid
  - Sets `grafana_session` cookie
  - If cookie is invalid `login.OAuthLogin(state mismatch)` error is returned 
  - If valid, navigate to default dashboard

Actual Issue:
Since the `oauth_state` cookie has only `60 seconds` as default expiry time, 
 - if the user takes more than that to enter the username or password when in login page 
 - loads the login page and come back after a break to login
 - signout from oauth provider navigates back to login page (as set in the config above), which sets the cookie and the user leaves the page for later login

All the above cases, user will get the same error. I can increase the cookie expiry time by setting `GF_AUTH_OAUTH_STATE_COOKIE_MAX_AGE` to desired value, but this issue can happen after this timeout as well.

Is this `oauth_state` check is required for any security purpose? If yes, is there a better way to handle the above problem without having to worry about cookie expiry time?

Note: Disabling oauth auto login also results in the same since, the cookie is set during {application_domain}/login api call

**Comment by marefr** [comment]
@gsaileshkumar `oauth_state` cookie is written when clicking on OAuth provider button from Grafana login page and right before being redirected to OAuth provider so disabling oauth auto login should make a difference as far as I know.

**Comment by gsaileshkumar** [comment]
@marefr Thanks for correcting. Yes verified that `oauth_state` cookie gets set in {application_domain}/login/generic_auth api call. So disabling oauth auto login solves the issue. 

However having auto login and redirecting to login page when logged out are two common user expectation for any application. Will there be an update for this issue in future?

**Comment by DanielMrowca** [comment]
I have the same issue but Only on CHROME ver 87.0.4280.88 and only after login from Identity Service
When i clear all parameters from url or click on "Grafana" everythink is ok

![obraz](https://user-images.githubusercontent.com/17824210/101185663-6edb0e00-3652-11eb-9bc9-69d5a6e57f86.png)

**Comment by oakad** [comment]
This issue still happens, though, and appears to be connected to the initial protocol choice in the negotiations. If the session starts with HTTP/1 it all works fine, but if the session sticks to HTTP/2 then various breakages ensue.

**Comment by waldemarennsaed** [comment]
Same issue on my side.
Using Grafana in a docker container in combination with keycloak, also in a docker-container.

Embedded grafana via `iframe` on a simple https - page. Then using SSO (keycloak) and that error is returning:

![Bildschirmfoto 2021-12-08 um 08 45 24](https://user-images.githubusercontent.com/94853611/145168808-ab217082-b540-4d20-8910-ad8cd0d337e0.png)

Embedded HTML:

```html
<iframe src="https://my-grafana-instance.com/login" sandbox="allow-scripts allow-same-origin allow-forms" frameborder="0" width="700" height="1080"></iframe>
```

Chrome itself only complains about the invalid cert (because I run a local HTTPS vue instance - the CERT shouldn't be the problem anyways):

![Bildschirmfoto 2021-12-08 um 08 47 00](https://user-images.githubusercontent.com/94853611/145169182-6908b2d4-b482-4ca1-ab00-beba8b4da61a.png)

The grafana logs **are not helpful at all** (first URL is the grafana domain, the others my localhost, IP of my device in the local network):

> grafana-instance    | t=2021-12-08T07:38:48+0000 lvl=eror msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/login/generic_oauth status=500 remote_addr=::ffff:195.14.214.162 time_ms=0 size=1754 referer=

Keycloak does not complain in any way, login is possible without any error logs.

**Comment by waldemarennsaed** [comment]
> we encountered the same issue, while the cause is not abount the `cookie_samesite` header.
> 
> It is because we are connecting to multiple grafana pages concurrently at the "same" time, and the concurrent connections are launced via iframe in the same host page. iframes would request a redirect link with diffrent oauth codes, of which only one will succeed:
> 
> ![image](https://user-images.githubusercontent.com/816618/137427273-1955a084-0cdd-4242-97e4-eade011c7126.png)
> 
> Our solution is simple: dont load iframes simultaneously, but load on need one by one.

Interesting. As stated above, I get the same logs in the browser (500 error from grafana) but we have only one auth-code process running at the time, which is the login via the iframe embedding. So I don't think that this is the origin problem for the issue, at least in my setup.

**Comment by meghanaph** [comment]
I have the same issue when I am using the latest keycloak version(19.0.0) for SSO and for the keyclaok with lower version(15.0.2) there is no such issue login was possible without any error logs

Once I click on sign in with oauth I am redirected with this error

"login.OAuthLogin(missing saved state)"

![grafana](https://user-images.githubusercontent.com/104889038/190323494-fd82ea4c-92d6-4998-b2b6-290dd450c2b3.jpg)

**Comment by nextMJ** [comment]
Having the same issue only when embedding via iframe. Multiple page refresh helps in the end but is very user-unfriendly. When it finally loads, it works like a charm until the next logout-login.

Having the latest version of Grafana (v9.2.5 (042e4d216b)) while using the latest Firefox (107.0). Running grafana with nginx proxy and using Keycloak as OAuth provider.

Grafana logs don't help at all as @waldemarennsaed pointed out. Tried all the config changes mentioned above without any success. `set-cookie` is present in the response packet.

![obrazek](https://user-images.githubusercontent.com/77353838/203380749-ade7cd52-2f07-4d6f-b767-ce172f8ccf95.png)

-> F5 -> F5 -> F5 -> F5 ->

![obrazek](https://user-images.githubusercontent.com/77353838/203381000-0bbb56c8-fe9d-4d7b-809a-2e10f8a931a4.png)

Can provide any information required because I'm lost here.

**Comment by shaheer-k** [comment]
@waldemarennsaed @nextMJ 
I am also facing the same issue _(login.OAuthLogin (missing saved state))_ when opening Grafana with keycloak login in iframe. Please let me know if you found any solution to it.

**Comment by rek** [comment]
Same here. Anyone resolve this?

**Comment by ethervoid** [comment]
Same here using Google OAuth and without iFrame

**Comment by msarti** [comment]
Same here using Azure AD OAuth2 and Grafana 9.3.6

**Comment by SirSkizo** [comment]
Same here using lastest version on granafa and keycloak. I have ensured that I have `cookie_samesite = lax`.

IMHO this ticket should be reopen

**Comment by guaj** [comment]
Same here, using Azure AD OAuth2 and Grafana 9.1.1. The SSO works like a charm when I open grafana in a new tab, but doesn't work inside an iFrame. 
Is there any way we can inject this https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/iframe-usage.md#apps-with-cross-origin into our grafana app ?

**Comment by bushong1** [comment]
I've been experiencing this with Firefox and no iframes, but I was able to resolve it by disabling browser enhanced tracking protection.
